src/hb/paint_extents.rs

```diff
@@ -135,62 +139,60 @@ impl<'a> hb_paint_extents_context_t<'a> {
         self.clips.pop();
     }
 
-    fn push_group(&mut self, mode: CompositeMode) {
-        let src_bounds = hb_bounds_t::default();
-
-        // In harfbuzz, this is in `pop_group` instead, but since we have the composite mode
-        // in push group, we need to do it here.
-        if let Some(backdrop_bounds) = self.groups.last_mut() {
-            match mode {
-                CompositeMode::Clear => backdrop_bounds.status = status_t::EMPTY,
-                CompositeMode::Source | CompositeMode::SourceOut => *backdrop_bounds = src_bounds,
-                CompositeMode::Destination | CompositeMode::DestinationOut => {}
-                CompositeMode::SourceIn | CompositeMode::DestinationIn => {
-                    if src_bounds.status == status_t::EMPTY {
-                        backdrop_bounds.status = status_t::EMPTY;
-                    } else if src_bounds.status == status_t::BOUNDED {
-                        backdrop_bounds.extents.x_min =
-                            backdrop_bounds.extents.x_min.max(src_bounds.extents.x_min);
-                        backdrop_bounds.extents.y_min =
-                            backdrop_bounds.extents.y_min.max(src_bounds.extents.y_min);
-                        backdrop_bounds.extents.x_max =
-                            backdrop_bounds.extents.x_max.min(src_bounds.extents.x_max);
-                        backdrop_bounds.extents.y_max =
-                            backdrop_bounds.extents.y_max.min(src_bounds.extents.y_max);
-
-                        if backdrop_bounds.extents.x_min >= backdrop_bounds.extents.x_max
-                            || backdrop_bounds.extents.y_min >= backdrop_bounds.extents.y_max
-                        {
-                            backdrop_bounds.status = status_t::EMPTY;
+    fn push_group(&mut self) {
+        self.groups.push(hb_bounds_t::default());
+    }
+
+    fn pop_group(&mut self) {
+        if let Some(mode) = self.composite_modes.pop() {
+            if let Some(src_bounds) = self.groups.pop() {
+                if let Some(backdrop_bounds) = self.groups.last_mut() {
+                    match mode {
+                        CompositeMode::Clear => backdrop_bounds.status = status_t::EMPTY,
+                        CompositeMode::Source | CompositeMode::SourceOut => *backdrop_bounds = src_bounds,
+                        CompositeMode::Destination | CompositeMode::DestinationOut => {}
+                        CompositeMode::SourceIn | CompositeMode::DestinationIn => {
+                            if src_bounds.status == status_t::EMPTY {
+                                backdrop_bounds.status = status_t::EMPTY;
+                            } else if src_bounds.status == status_t::BOUNDED {
+                                backdrop_bounds.extents.x_min =
```

Comment:
How about a wrapper method for this? Like `hb_bounds_t::extend(&mut self, other: hb_bounds_t)`

Reply:
Immediately outdated, nice!